File: Item.js
const { DataTypes } = require('sequelize');
const { sequelize } = require('../config/database');
const { v4: uuidv4 } = require('uuid');

const Item = sequelize.define('Item', {
  id: {
    type: DataTypes.UUID,
    defaultValue: DataTypes.UUIDV4,
    primaryKey: true
  },
  itemCode: {
    type: DataTypes.STRING,
    unique: true,
    allowNull: false,
    comment: 'Código único do item para QR Code'
  },
  name: {
    type: DataTypes.STRING,
    allowNull: false
  },
  description: {
    type: DataTypes.TEXT
  },
  categoryId: {
    type: DataTypes.UUID,
    allowNull: false,
    references: {
      model: 'categories',
      key: 'id'
    }
  },
  locationId: {
    type: DataTypes.UUID,
    references: {
      model: 'locations',
      key: 'id'
    }
  },
  quantity: {
    type: DataTypes.INTEGER,
    defaultValue: 0,
    validate: {
      min: 0
    }
  },
  minQuantity: {
    type: DataTypes.INTEGER,
    defaultValue: 1,
    comment: 'Quantidade mínima para alerta de estoque baixo'
  },
  maxQuantity: {
    type: DataTypes.INTEGER,
    comment: 'Quantidade máxima permitida'
  },
  unit: {
    type: DataTypes.STRING,
    defaultValue: 'un',
    comment: 'Unidade de medida (un, kg, m, L, etc)'
  },
  barcode: {
    type: DataTypes.STRING,
    comment: 'Código de barras do fornecedor (se houver)'
  },
  serialNumber: {
    type: DataTypes.STRING,
    comment: 'Número de série (para equipamentos)'
  },
  manufacturer: {
    type: DataTypes.STRING,
    comment: 'Fabricante'
  },
  model: {
    type: DataTypes.STRING,
    comment: 'Modelo'
  },
  acquisitionDate: {
    type: DataTypes.DATE,
    comment: 'Data de aquisição'
  },
  expirationDate: {
    type: DataTypes.DATE,
    comment: 'Data de validade (para consumíveis)'
  },
  price: {
    type: DataTypes.DECIMAL(10, 2),
    comment: 'Preço unitário'
  },
  condition: {
    type: DataTypes.ENUM('new', 'good', 'fair', 'poor', 'damaged'),
    defaultValue: 'good'
  },
  status: {
    type: DataTypes.ENUM('available', 'in_use', 'maintenance', 'reserved', 'retired'),
    defaultValue: 'available'
  },
  currentUserId: {
    type: DataTypes.UUID,
    references: {
      model: 'users',
      key: 'id'
    },
    comment: 'Usuário atual que está usando o item'
  },
  image: {
    type: DataTypes.STRING,
    comment: 'URL da imagem do item'
  },
  qrCodeUrl: {
    type: DataTypes.STRING,
    comment: 'URL da imagem do QR Code gerado'
  },
  notes: {
    type: DataTypes.TEXT
  },
  tags: {
    type: DataTypes.ARRAY(DataTypes.STRING),
    defaultValue: []
  },
  isActive: {
    type: DataTypes.BOOLEAN,
    defaultValue: true
  }
}, {
  tableName: 'items',
  timestamps: true,
  hooks: {
    beforeCreate: (item) => {
      // Gerar código único se não fornecido
      if (!item.itemCode) {
        const prefix = item.name.substring(0, 3).toUpperCase();
        const timestamp = Date.now().toString(36).toUpperCase();
        const random = Math.random().toString(36).substring(2, 5).toUpperCase();
        item.itemCode = `${prefix}-${timestamp}-${random}`;
      }
    }
  }
});

module.exports = Item;

File: item.dart
import 'package:hive/hive.dart';

part 'item.g.dart';

@HiveType(typeId: 0)
class Item {
  @HiveField(0)
  final String id;

  @HiveField(1)
  final String itemCode;

  @HiveField(2)
  final String name;

  @HiveField(3)
  final String? description;

  @HiveField(4)
  final String categoryId;

  @HiveField(5)
  final String categoryName;

  @HiveField(6)
  final String? locationId;

  @HiveField(7)
  final String? locationCode;

  @HiveField(8)
  final int quantity;

  @HiveField(9)
  final int minQuantity;

  @HiveField(10)
  final String unit;

  @HiveField(11)
  final String? serialNumber;

  @HiveField(12)
  final String? manufacturer;

  @HiveField(13)
  final String? model;

  @HiveField(14)
  final String condition;

  @HiveField(15)
  final String status;

  @HiveField(16)
  final String? currentUserId;

  @HiveField(17)
  final String? image;

  @HiveField(18)
  final String? qrCodeUrl;

  @HiveField(19)
  final List<String> tags;

  @HiveField(20)
  final DateTime createdAt;

  @HiveField(21)
  final DateTime updatedAt;

  @HiveField(22)
  final bool isSynced;

  Item({
    required this.id,
    required this.itemCode,
    required this.name,
    this.description,
    required this.categoryId,
    required this.categoryName,
    this.locationId,
    this.locationCode,
    required this.quantity,
    required this.minQuantity,
    required this.unit,
    this.serialNumber,
    this.manufacturer,
    this.model,
    required this.condition,
    required this.status,
    this.currentUserId,
    this.image,
    this.qrCodeUrl,
    required this.tags,
    required this.createdAt,
    required this.updatedAt,
    this.isSynced = true,
  });

  factory Item.fromJson(Map<String, dynamic> json) {
    return Item(
      id: json['id'],
      itemCode: json['itemCode'],
      name: json['name'],
      description: json['description'],
      categoryId: json['categoryId'],
      categoryName: json['category']?['name'] ?? '',
      locationId: json['locationId'],
      locationCode: json['location']?['coordinates'],
      quantity: json['quantity'] ?? 0,
      minQuantity: json['minQuantity'] ?? 1,
      unit: json['unit'] ?? 'un',
      serialNumber: json['serialNumber'],
      manufacturer: json['manufacturer'],
      model: json['model'],
      condition: json['condition'] ?? 'good',
      status: json['status'] ?? 'available',
      currentUserId: json['currentUserId'],
      image: json['image'],
      qrCodeUrl: json['qrCodeUrl'],
      tags: List<String>.from(json['tags'] ?? []),
      createdAt: DateTime.parse(json['createdAt']),
      updatedAt: DateTime.parse(json['updatedAt']),
      isSynced: true,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'itemCode': itemCode,
      'name': name,
      'description': description,
      'categoryId': categoryId,
      'locationId': locationId,
      'quantity': quantity,
      'minQuantity': minQuantity,
      'unit': unit,
      'serialNumber': serialNumber,
      'manufacturer': manufacturer,
      'model': model,
      'condition': condition,
      'status': status,
      'currentUserId': currentUserId,
      'image': image,
      'tags': tags,
    };
  }

  Item copyWith({
    String? id,
    String? itemCode,
    String? name,
    String? description,
    String? categoryId,
    String? categoryName,
    String? locationId,
    String? locationCode,
    int? quantity,
    int? minQuantity,
    String? unit,
    String? serialNumber,
    String? manufacturer,
    String? model,
    String? condition,
    String? status,
    String? currentUserId,
    String? image,
    String? qrCodeUrl,
    List<String>? tags,
    DateTime? createdAt,
    DateTime? updatedAt,
    bool? isSynced,
  }) {
    return Item(
      id: id ?? this.id,
      itemCode: itemCode ?? this.itemCode,
      name: name ?? this.name,
      description: description ?? this.description,
      categoryId: categoryId ?? this.categoryId,
      categoryName: categoryName ?? this.categoryName,
      locationId: locationId ?? this.locationId,
      locationCode: locationCode ?? this.locationCode,
      quantity: quantity ?? this.quantity,
      minQuantity: minQuantity ?? this.minQuantity,
      unit: unit ?? this.unit,
      serialNumber: serialNumber ?? this.serialNumber,
      manufacturer: manufacturer ?? this.manufacturer,
      model: model ?? this.model,
      condition: condition ?? this.condition,
      status: status ?? this.status,
      currentUserId: currentUserId ?? this.currentUserId,
      image: image ?? this.image,
      qrCodeUrl: qrCodeUrl ?? this.qrCodeUrl,
      tags: tags ?? this.tags,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      isSynced: isSynced ?? this.isSynced,
    );
  }

  bool get isLowStock => quantity <= minQuantity;
  bool get isAvailable => status == 'available' && quantity > 0;
}
